packages/core/src/mixer.ts

```diff
@@ -25,42 +23,129 @@ const comparator = (a: HeapElement, b: HeapElement): number => {
  */
 export class Mixer<Chain extends HoprCoreConnector> {
   private queue: Heap<HeapElement>
+  private timeout?: NodeJS.Timeout
+  private defer: DeferredPromise<void>
+  private deferEnd: DeferredPromise<void>
+  private endPromise: Promise<void>
+  private _done: boolean
 
   public WAIT_TIME = MAX_PACKET_DELAY
 
   constructor(private incrementer = Date.now) {
     this.queue = new Heap(comparator)
-  }
 
-  push(p: Packet<Chain>) {
-    this.queue.push([this.getPriority(), p])
+    this.defer = Defer<void>()
+    this.deferEnd = Defer<void>()
+
+    this._done = false
+
+    this.endPromise = this.deferEnd.promise.then(() => {
+      this._done = true
+    })
   }
 
-  // Can we pop an element?.
-  poppable(): boolean {
-    log(`Mixer has ${this.queue.length} elements`)
-    if (!this.queue.length) {
-      return false
+  /**
+   * Adds a packet to the mixer
+   * @param p
+   */
+  public push(p: Packet<Chain>) {
+    if (this._done) {
+      throw Error(`Mixer has ended. Could not accept any further messages.`)
     }
-    return this.queue.peek()[0] < this.due()
-  }
+    const newPriority = this.getPriority()
+    const topPriority = this.queue.length > 0 ? this.queue.peek()[0] : Number.MAX_SAFE_INTEGER
```

Comment:
This is less readable than before

Reply:
Absolutely, because we need to know both values to determine whether we need to reset the timeout. This was not necessary before.

Reply:
we don't need MAX_INTEGER - if the queue length is 0 we can just reset the timeout as well.

Reply:
I think we have reached the discussion about personal coding style here

Reply:
removed MAX_SAFE_INTEGER